src/random.cpp

```diff
@@ -272,71 +301,137 @@ void GetOSRand(unsigned char *ent32)
-    hasher.Write((const unsigned char*)&len, sizeof(len));
-    hasher.Write((const unsigned char*)data, len);
-    unsigned char buf[64];
+    // OpenSSL RNG (for now)
     {
-        WAIT_LOCK(cs_rng_state, lock);
-        hasher.Write(rng_state, sizeof(rng_state));
-        hasher.Write((const unsigned char*)&rng_counter, sizeof(rng_counter));
-        ++rng_counter;
-        hasher.Finalize(buf);
-        memcpy(rng_state, buf + 32, 32);
+        std::lock_guard<std::mutex> lock(g_cs_openssl);
+        RAND_bytes(buffer, sizeof(buffer));
+    }
+    hasher.Write(buffer, sizeof(buffer));
+
+    // High-precision timestamp.
+    //
+    // Note that we also commit to a timestamp in the Fast seeder, so we indirectly commit to a
+    // benchmark of all the entropy gathering sources in this function).
+    SeedTimestamp(hasher);
+}
+
+static void SeedSlow(CSHA512& hasher)
+{
+    // Everything that the 'fast' and 'medium' seeders include.
+    SeedMedium(hasher);
+
+    // Sleep for 1ms
+    MilliSleep(1);
+
+    // High-precision timestamp after sleeping
+    int64_t perfcounter = GetPerformanceCounter();
+    hasher.Write((const unsigned char*)&perfcounter, sizeof(perfcounter));
+
+    // Windows performance monitor data (once every 10 minutes)
+    RandAddSeedPerfmon(hasher);
+
+    // Once per minute, strengthen the hash for 10ms.
+    static int64_t last_strengthen;
+    int64_t current_time = GetTimeMicros();
+    if (current_time > last_strengthen + 60000) {
+        // Write randomness from the actual RNG in the state, so it goes through the strengthening as well.
+        unsigned char buffer[32];
+        GetRandBytes(buffer, 32);
+        hasher.Write(buffer, 32);
+
+        Strengthen(hasher, 10000);
+        last_strengthen = current_time;
+    }
+}
+
+static void SeedStartup(CSHA512& hasher)
+{
+#ifdef WIN32
+    {
+        std::lock_guard<std::mutex> lock(g_cs_openssl);
+        RAND_screen();
     }
-    memory_cleanse(buf, 64);
+#endif
+
+    // Everything that the 'fast', 'medium', and 'slow' seeders include.
+    SeedSlow(hasher);
 }
 
-void GetStrongRandBytes(unsigned char* out, int num)
+static std::once_flag g_rng_initialized;
+static Mutex cs_rng_state;
+static unsigned char rng_state[32] GUARDED_BY(cs_rng_state) = {0};
+static uint64_t rng_counter GUARDED_BY(cs_rng_state) = 0;
+
+enum class RNGLevel {
+    FAST, //!< Automatically called by GetRandBytes
+    MEDIUM, //!< Automatically called by GetStrongRandBytes
+    SLOW, //!< Called by RandSeed() and at startup
+    STARTUP, //!< Called only at startup.
+};
+
+static void ProcRand(unsigned char* out, int num, RNGLevel level)
 {
     assert(num <= 32);
-    CSHA512 hasher;
-    unsigned char buf[64];
 
-    // First source: OpenSSL's RNG
-    RandAddSeedPerfmon();
-    GetRandBytes(buf, 32);
-    hasher.Write(buf, 32);
+    std::call_once(g_rng_initialized, [&](){
+        // The first invocation of the RNG always gathers all we have.
+        level = RNGLevel::STARTUP;
```

Comment:
Very minor nit: Likely a matter of taste, but personally I try to avoid parameter reassignment. Would it be possible to avoid parameter re-assignment and make the parameter `level` const? Feel free to ignore this suggestion.